gtsam/sfm/DsfTrackGenerator.h

```diff
@@ -0,0 +1,207 @@
+using MatchIndicesMap = std::map<IndexPair, CorrespondenceIndices>;
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement {
+  size_t i;
+  gtsam::Point2 uv;
+
+  NamedSfmMeasurement(size_t i, gtsam::Point2 uv) : i(i), uv(uv) {}
+};
+
+
+/**
+ * @brief Track containing 2D measurements associated with a single 3D point.
+ * Note: Equivalent to gtsam.SfmTrack, but without the 3d measurement.
+ * This class holds data temporarily before 3D point is initialized.
+ */
+class SfmTrack2d {
+ private:
+  std::vector<NamedSfmMeasurement> measurements_;
+
+ public:
+  // Default constructor.
+  SfmTrack2d() = default;
+
+  // Constructor from measurements.
+  SfmTrack2d(std::vector<NamedSfmMeasurement> &measurements) : measurements_(measurements) {}
+
+  // Add a measurement to the track.
+  void addMeasurement(const NamedSfmMeasurement &m) {
+    measurements_.emplace_back(m);
+  }
+
+  /// The measurement at index `idx`
+  NamedSfmMeasurement measurement(size_t idx) const { return measurements_[idx]; }
+
+  // Return all measurements in the track.
+  std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
+
+  /// Total number of measurements in this track.
+  size_t numberMeasurements() const { return measurements_.size(); }
+
+  // @brief Validates the track by checking that no two measurements are from the same camera.
+  //
+  // returns boolean result of the validation.
+  bool hasUniqueCameras()
+  {
+    std::vector<int> track_cam_idxs;
+    for (auto & measurement : measurements_)
+    {
+      track_cam_idxs.emplace_back(measurement.i);
+    }
+    auto i = std::adjacent_find(track_cam_idxs.begin(), track_cam_idxs.end());
+    bool all_cameras_unique = (i == track_cam_idxs.end());
+    return all_cameras_unique;
+  }
+};
+
+using SfmTrack2dVector = std::vector<gtsam::SfmTrack2d>;
+using NamedSfmMeasurementVector = std::vector<NamedSfmMeasurement>;
+
+
+/**
+ * @brief Generates point tracks from connected components in the keypoint matches graph.
+ */
+class DsfTrackGenerator {
+
+ public:
+  /** Default constructor. */
+  DsfTrackGenerator() {}
+
+  // Creates a list of tracks from 2d point correspondences.
+  //
+  // Creates a disjoint-set forest (DSF) and 2d tracks from pairwise matches.
+  // We create a singleton for union-find set elements from camera index of a
+  // detection and the index of that detection in that camera's keypoint list,
+  // i.e. (i,k).
+  // @param Map from (i1,i2) image pair indices to (K,2) matrix, for K
+  //        correspondence indices, from each image.
+  // @param Length-N list of keypoints, for N images/cameras.
+  std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+    const MatchIndicesMap& matches_dict,
+    const KeypointsList& keypoints_list) {
+    std::vector<SfmTrack2d> track_2d_list;
+
+    std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
```

Comment:
Do we need this?

Reply:
Maybe only do it if a verbose flag is set ?

Reply:
Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.